modules/blocks/README.md

```diff
@@ -0,0 +1,162 @@
+
+- There's no built-in JavaScript MD5 hash function, so we pull in [a common implementation](https://github.com/blueimp/JavaScript-MD5).
+- When registering a new block, you must prefix its slug with a namespace for your plugin. This helps prevent conflicts when more than one plugin registers a block with the same slug.
+- You will use `createElement` to describe the structure of your block's markup. See the [Element documentation](../element/README.md) for more information.
+- Extracting `GravatarImage` to a separate function allows us to reuse it in both the editor-specific interface and the published content.
+- We only change the attributes of a block by calling the `setAttributes` helper. Never assign a value on the attributes object directly.
+
+There's a few small details that aren't obvious though:
+
+- How are a block's attributes stored when saving a post, so that they're restored next time I edit the same post?
+- Keen to avoid spam, how can I ensure I'm not including the email address in the published content?
+
+The single answer to both of these questions is that the combined state of every block in a post is saved as [post meta](https://codex.wordpress.org/Custom_Fields#Advanced_Techniques_for_Custom_Fields). This ensures that the attributes of a block are preserved and kept separate from the markup between editing sessions. 
+
+## API
+
+### `wp.blocks.registerBlock( slug: string, settings: Object )`
+
+Registers a new block provided a unique slug and an object defining its behavior. Once registered, the block is made available as an option to any editor interface where blocks are implemented.
+
+- `name: string` - A human-readable [localized](https://codex.wordpress.org/I18n_for_WordPress_Developers#Handling_JavaScript_files) label for the block. Shown in the block picker.
+- `edit( { attributes: Object, setAttributes: Function } ): WPElement` - Returns an element describing the markup of a block to be shown in the editor. A block can update its own state in response to events using the `setAttributes` function, passing an object of properties to be applied as a partial update.
+- `save( { attributes: Object } ): WPElement` - Returns an element describing the markup of a block to be saved in the published content. This function is called before save and when switching to an editor's HTML view.
+- `tagName: string` - An alternative to defining `edit` and `save` behaviors, if passed a [tag name](https://developer.mozilla.org/en-US/docs/Web/API/Document/createElement#Parameters), delegates default behavior of an editable field on that node type. Typically used for simple, text-heavy elements (paragraphs, headings) to avoid excess meta storage of text.
+- `controls: string[]` - Slugs for controls to be made available to block. See also: [`wp.blocks.registerControl`](#wpblocksregistercontrol-slug-string-settings-object-)
+
+### `wp.blocks.registerControl( slug: string, settings: Object )`
```

Comment:
It's a good point to raise. A few from the mockups that might be hard to implement with this direction:

- Linking: Presumably need to be able to render some tooltip. `setAttributes` of the `onClick` callback can at least be asynchronous from current API proposal. Some general `render`-like behavior could be a good fit. But in that case would a control need its own state (shown/hidden)?
- Floating toolbars on "Image in gallery selected"
- Distinct toolbar buckets, like "Early Admin UI Concept" image advanced options button. This would be much simpler if we simplify distinction between block-level and inline controls (advanced options as button grouped with rest of alignment). 

Some related Slack discussion: https://wordpress.slack.com/archives/C02QB2JS7/p1490119310980240